File: platform/statistics/devkit/src/com/intellij/internal/statistic/actions/AddTestGroupToLocalWhitelistAction.java
// Copyright 2000-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.internal.statistic.actions;

import com.intellij.icons.AllIcons;
import com.intellij.idea.ActionsBundle;
import com.intellij.internal.statistic.StatisticsBundle;
import com.intellij.internal.statistic.eventLog.whitelist.WhitelistTestGroupStorage;
import com.intellij.notification.Notification;
import com.intellij.notification.NotificationType;
import com.intellij.notification.Notifications;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.ui.LayeredIcon;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

/**
 * Action adds not whitelisted statistics group to a local whitelist for testing.
 * <p>
 * If "Add custom validation rules" is disabled, all event id and event data values from the group will be allowed.
 */
public class AddTestGroupToLocalWhitelistAction extends AnAction {
  private static final LayeredIcon ICON = new LayeredIcon(AllIcons.General.Add, AllIcons.Actions.Scratch);

  public AddTestGroupToLocalWhitelistAction() {
    super(ActionsBundle.message("action.AddTestGroupToLocalWhitelistAction.text"),
          ActionsBundle.message("action.AddTestGroupToLocalWhitelistAction.description"),
          ICON);
  }

  @Override
  public void actionPerformed(@NotNull AnActionEvent e) {
    final Project project = e.getProject();
    if (project == null) {
      return;
    }

    final AddGroupToLocalWhitelistDialog dialog = new AddGroupToLocalWhitelistDialog(project, null, null);
    final boolean result = dialog.showAndGet();
    if (!result || StringUtil.isEmpty(dialog.getGroupId()) || StringUtil.isEmpty(dialog.getRecorderId())) {
      return;
    }

    ProgressManager.getInstance().run(new Task.Backgroundable(project, "Adding Test Group and Updating Whitelist...", false) {
      @Override
      public void run(@NotNull ProgressIndicator indicator) {
        final String recorderId = dialog.getRecorderId();
        final WhitelistTestGroupStorage testWhitelist = WhitelistTestGroupStorage.getInstance(recorderId);
        try {
          if (dialog.isCustomRules()) {
            testWhitelist.addGroupWithCustomRules(dialog.getGroupId(), dialog.getCustomRules());
          }
          else {
            testWhitelist.addTestGroup(dialog.getGroupId());
          }
          showNotification(project, NotificationType.INFORMATION, "Group '" + dialog.getGroupId() + "' was added to local whitelist");
        }
        catch (IOException ex) {
          showNotification(project, NotificationType.ERROR, "Failed updating local list: " + ex.getMessage());
        }
      }
    });
  }

  protected void showNotification(@NotNull Project project,
                                  @NotNull NotificationType type,
                                  @NotNull String message) {
    String title = StatisticsBundle.message("stats.feature.usage.statistics");
    Notifications.Bus.notify(new Notification("FeatureUsageStatistics", title, message, type), project);
  }
}

File: platform/statistics/devkit/src/com/intellij/internal/statistic/actions/localWhitelist/AddTestGroupToLocalWhitelistAction.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.internal.statistic.actions.localWhitelist

import com.intellij.icons.AllIcons
import com.intellij.idea.ActionsBundle
import com.intellij.internal.statistic.StatisticsBundle
import com.intellij.internal.statistic.StatisticsDevKitUtil
import com.intellij.internal.statistic.StatisticsDevKitUtil.showNotification
import com.intellij.internal.statistic.eventLog.whitelist.LocalWhitelistGroup
import com.intellij.internal.statistic.eventLog.whitelist.WhitelistTestGroupStorage
import com.intellij.internal.statistic.service.fus.FUStatisticsWhiteListGroupsService
import com.intellij.notification.NotificationType
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.progress.runBackgroundableTask
import com.intellij.openapi.project.DumbAwareAction
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.util.Disposer
import com.intellij.ui.components.dialog
import java.io.IOException

/**
 * Action adds not whitelisted statistics group to a local whitelist for testing.
 *
 * If "Add custom validation rules" is disabled, all event id and event data values from the group will be allowed.
 */
class AddTestGroupToLocalWhitelistAction constructor(private val recorderId: String = StatisticsDevKitUtil.DEFAULT_RECORDER)
  : DumbAwareAction(ActionsBundle.messagePointer("action.AddTestGroupToLocalWhitelistAction.text"),
                    ActionsBundle.messagePointer("action.AddTestGroupToLocalWhitelistAction.description"),
                    AllIcons.General.Add) {

  override fun actionPerformed(e: AnActionEvent) {
    val project = e.project ?: return
    val testWhitelistStorage = WhitelistTestGroupStorage.getTestStorage(recorderId)
    if (testWhitelistStorage == null) {
      showNotification(project, NotificationType.ERROR, StatisticsBundle.message("stats.cannot.find.test.whitelist.storage"))
      return
    }
    val whitelistGroup = LocalWhitelistGroup("", false)
    val dialog = createAddToLocalWhitelistDialog(project, testWhitelistStorage, whitelistGroup)
    if (dialog == null || !dialog.showAndGet()) return

    runBackgroundableTask(StatisticsBundle.message("stats.updating.local.whitelist"), project, false) {
      try {
        testWhitelistStorage.addTestGroup(whitelistGroup)
        showNotification(project, NotificationType.INFORMATION, StatisticsBundle.message("stats.local.whitelist.was.updated"))
      }
      catch (ex: IOException) {
        showNotification(project, NotificationType.ERROR, StatisticsBundle.message("stats.failed.updating.local.whitelist.0", ex.message))
      }
    }
  }

  companion object {
    fun createAddToLocalWhitelistDialog(project: Project,
                                        testGroupStorage: WhitelistTestGroupStorage,
                                        group: LocalWhitelistGroup): DialogWrapper? {
      val productionGroups = loadProductionGroups(project, testGroupStorage)
      if (productionGroups == null) return null
      val groupConfiguration = LocalWhitelistGroupConfiguration(project, productionGroups, group)
      val dialog = dialog(
        StatisticsBundle.message("stats.add.test.group.to.local.whitelist"),
        panel = groupConfiguration.panel,
        resizable = true,
        focusedComponent = groupConfiguration.getFocusedComponent(),
        project = project,
        ok = { groupConfiguration.validate() }
      )
      Disposer.register(dialog.disposable, groupConfiguration)
      return dialog
    }

    private fun loadProductionGroups(project: Project,
                                     testGroupStorage: WhitelistTestGroupStorage): FUStatisticsWhiteListGroupsService.WLGroups? {
      return ProgressManager.getInstance().run(object : Task.WithResult<FUStatisticsWhiteListGroupsService.WLGroups?, IOException>(
        project, StatisticsBundle.message("stats.loading.validation.rules"), true) {
        override fun compute(indicator: ProgressIndicator): FUStatisticsWhiteListGroupsService.WLGroups? {
          val productionGroups = testGroupStorage.loadProductionGroups()
          if (indicator.isCanceled) return null
          return productionGroups
        }
      })
    }
  }

}